-feat(Sprint 14(1)):

ArsenyIshtubaev opened the pull request:
-добавить работу с базой данных
-дать пользователям возможность бронировать вещи

ArsenyIshtubaev committed:
-feat(Sprint 14(1)):
-добавить работу с базой данных
-дать пользователям возможность бронировать вещи

ArsenyIshtubaev committed:
-feat(Sprint 14(1)):
-добавить работу с базой данных
-дать пользователям возможность бронировать вещи

ArsenyIshtubaev committed:
-feat(Sprint 14(1)):
-добавить работу с базой данных
-дать пользователям возможность бронировать вещи

ArsenyIshtubaev committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

ArsenyIshtubaev committed:
-feat(Sprint 14(1)):
-добавить работу с базой данных
-дать пользователям возможность бронировать вещи

DyndaNE commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+        List<BookingDto> result = bookingRepository.findAll().stream().filter(booking ->
+                booking.getBooker().getId() == (userId)).map(mapper::toBookingDto).collect(Collectors.toList());
+        switch (state) {
+            case ("ALL"):
```
молодец, что использовал switch, но почему не используешь enum?

DyndaNE commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+            throw new StorageException("Incorrect userId");
+        }
+        List<BookingDto> result = bookingRepository.findAll().stream().filter(booking ->
+                booking.getBooker().getId() == (userId)).map(mapper::toBookingDto).collect(Collectors.toList());
```
не стоит так делать, получать все,А потом фильтровать, это очень долго будет работать. Стоит делать отдельный запрос для каждого состояния

DyndaNE commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+        }
+        Booking booking = mapper.fromSimpleToBooking(bookingDtoSimple);
+        booking.setBooker(userRepository.findById(userId).orElseThrow());
+        if (itemRepository.findById(bookingDtoSimple.getItemId()).isEmpty()) {
```
зачем делать обращение в репозиторий каждый раз? выполни один раз, сохрани в переменную и проверяй её потом любым способом

DyndaNE commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+            throw new StorageException("У пользователя нет вещей");
+        }
+        switch (state) {
+            case ("ALL"):
```
тут тоже стоит использовать enum

DyndaNE commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -3,22 +3,35 @@
-    private ItemRequest request;
+    /*@ManyToOne(optional=false)
+    @JoinColumn(name="request_id")
+    private ItemRequest request; */
```
почему тут так странно закомментировано? стоит его вообще удалить, тут не должно быть этого поля

ArsenyIshtubaev committed:
-fix (Sprint 14(2)):
-добавить enum в BookingServiceImpl
-сделать отдельный запрос в БД для каждого состояния state

DyndaNE commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,199 @@
+                    && bookingRepository.findById(bookingId).get().getItem().getOwner().getId() != userId) {
+                throw new StorageException("Incorrect userId");
+            }
+            return mapper.toBookingDto(bookingRepository.findById(bookingId).get());
```
ты в этом методе делаешь 4 вызова  bookingRepository.findById(bookingId), тут осталась старая ошибка о которой я писала в прошлый раз

DyndaNE commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -20,28 +29,80 @@ public class ItemServiceImpl implements ItemService {
+                    .toItemDtoWithBooking(itemRepository.findById(itemId).get());
+            if (itemRepository.findById(itemId).get().getOwner().getId() == userId) {
+                createItemDtoWithBooking(itemDtoWithBooking);
+            }
```
Арсений, стоит продумывать методы оптимальнее, обрати внимание сколько раз тут вызывается itemRepository.findById(itemId)

DyndaNE commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -20,28 +29,80 @@ public class ItemServiceImpl implements ItemService {
+            createItemDtoWithBooking(itemDtoWithBooking);
+            if (!commentRepository.findAllByItem_Id(itemDtoWithBooking.getId()).isEmpty()) {
+                itemDtoWithBooking.setComments(commentRepository.findAllByItem_Id(itemDtoWithBooking.getId())
+                        .stream().map(commentMapper::toCommentDto)
```
и тут  commentRepository.findAllByItem_Id(itemDtoWithBooking.getId() каждый запрос вызывает замедление работы

DyndaNE commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -20,28 +29,80 @@ public class ItemServiceImpl implements ItemService {
+                        LocalDateTime.now()).isEmpty()) {
+            BookingDtoForItem nextBooking = bookingMapper.toBookingDtoForItem(bookingRepository
+                    .findBookingsByItem_IdAndStartIsAfterOrderByStartDesc(itemDtoWithBooking.getId(),
+                            LocalDateTime.now()).get(0));
```
тут тоже вызываешь просто так дважды обращение в базу,  что один что другой метод

ArsenyIshtubaev committed:
-fix (Sprint 14(3)):
-оптимизировать время выполнения методов сервисов

DyndaNE commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -20,28 +31,78 @@ public class ItemServiceImpl implements ItemService {
+                        .stream().map(commentMapper::toCommentDto)
+                        .collect(Collectors.toList()));
+            }
+        }
```
молодец, я тебя наверное этим замучала, но так намного лучше. Всегда обращай внимание если ты запрашиваешь что то не один раз, то выноси в переменную.
если ты постоянно получаешь какое то поле из дто его тоже стоит вынести в переменную. и использовать её.

ArsenyIshtubaev replied:
Привет! Все отлично. Спасибо, за твою работу. Замечания всегда по делу.
